src/object.d

```diff
@@ -3241,6 +3241,226 @@ template RTInfo(T)
     enum RTInfo = null;
 }
 
+int _adCmpT(T1, T2)(T1[] s1, T2[] s2)
+{
+    // U1 gets the unqualified version of T1
+    static if (is(T1 W == immutable W)) alias U1 = W;
+    else static if (is(T1 W == const W)) alias U1 = W;
+    else alias U1 = T1;
+
+    // U2 gets the unqualified version of T2
+    static if (is(T2 U == immutable U)) alias U2 = U;
+    else static if (is(T2 U == const U)) alias U2 = U;
+    else alias U2 = T2;
+
+    // objects
+    static if (is(U1 : Object) && is(U2 : Object))
+    {
+        auto c  = cast(sizediff_t)(s1.length - s2.length);
+        if (c == 0)
```

Comment:
It looks like I was wrong. The following code:
```
int cmp(Object[] a, Object[] b)
{
    return a < b;
}
```
generates:
```
_D18TypeInfo_AC6Object6__initZ  comdat
        dd      offset FLAT:_D14TypeInfo_Array6__vtblZ
        db      000h,000h,000h,000h     ;....
        dd      offset FLAT:_D6Object7__ClassZ

_D3bug3cmpFAC6ObjectAC6ObjectZi comdat
        assume  CS:_D3bug3cmpFAC6ObjectAC6ObjectZi
L0:             mov     EAX,offset FLAT:_D18TypeInfo_AC6Object6__initZ
                push    EAX
                push    dword ptr 0Ch[ESP]
                push    dword ptr 0Ch[ESP]
                push    dword ptr 01Ch[ESP]
                push    dword ptr 01Ch[ESP]
                call    near ptr __adCmp2
                shr     EAX,01Fh
                add     ESP,014h
                ret     010h
```
meaning it is calling `TypeInfo_Array` rather than `TypeInfo_AC`. The latter is never called any more that I can see. I don't know when this changed, it's been a long time since I've looked at it. This dead code should be removed from druntime. The same goes for `TypeInfo_Ag`.

The code to generate a typeinfo looks like this, in typinf.d:
```
extern (C++) TypeInfoDeclaration getTypeInfoDeclaration(Type t)
{
    //printf("Type::getTypeInfoDeclaration() %s\n", t.toChars());
    switch (t.ty)
    {
    case Tpointer:
        return TypeInfoPointerDeclaration.create(t);
    case Tarray:
        return TypeInfoArrayDeclaration.create(t);
    case Tsarray:
        return TypeInfoStaticArrayDeclaration.create(t);
    case Taarray:
        return TypeInfoAssociativeArrayDeclaration.create(t);
    case Tstruct:
        return TypeInfoStructDeclaration.create(t);
    case Tvector:
        return TypeInfoVectorDeclaration.create(t);
    case Tenum:
        return TypeInfoEnumDeclaration.create(t);
    case Tfunction:
        return TypeInfoFunctionDeclaration.create(t);
    case Tdelegate:
        return TypeInfoDelegateDeclaration.create(t);
    case Ttuple:
        return TypeInfoTupleDeclaration.create(t);
    case Tclass:
        if ((cast(TypeClass)t).sym.isInterfaceDeclaration())
            return TypeInfoInterfaceDeclaration.create(t);
        else
            return TypeInfoClassDeclaration.create(t);

    default:
        return TypeInfoDeclaration.create(t);
    }
}
```
and from todt.d:
```
    override void visit(TypeInfoArrayDeclaration d)
    {
        //printf("TypeInfoArrayDeclaration.toDt()\n");
        verifyStructSize(Type.typeinfoarray, 3 * Target.ptrsize);

        dtb.xoff(toVtblSymbol(Type.typeinfoarray), 0);    // vtbl for TypeInfo_Array
        dtb.size(0);                                     // monitor

        assert(d.tinfo.ty == Tarray);

        TypeDArray tc = cast(TypeDArray)d.tinfo;

        genTypeInfo(tc.next, null);
        dtb.xoff(toSymbol(tc.next.vtinfo), 0); // TypeInfo for array of type
    }
```
where there are no special cases for certain array types.